File: leetcode/easy/3105-longest-strictly-increasing-or-strictly-decreasing-subarray/3105-longest-strictly-increasing-or-strictly-decreasing-subarray.kt
// 3105-longest-strictly-increasing-or-strictly-decreasing-subarray
// URL: https://leetcode.com/problems/longest-strictly-increasing-or-strictly-decreasing-subarray/description/


class Solution {
    fun max(nums: IntArray): Int {
        var maxValue = nums[0]

        for (i in 1 until nums.size) {
            if (nums[i] > maxValue) {
                maxValue = nums[i]
            }
        }

        return maxValue
    }

    fun longestMonotonicSubarray(nums: IntArray): Int {
        var longestMonotonicSubarrayLength = 1
        var increaseConsecutiveCount = 1
        var decreaseConsecutiveCount = 1

        for (i in 1 until nums.size) {
            if (nums[i] > nums[i - 1]) {
                increaseConsecutiveCount++
                decreaseConsecutiveCount = 1
            } else if (nums[i] < nums[i - 1]) {
                decreaseConsecutiveCount++
                increaseConsecutiveCount = 1
            } else {
                increaseConsecutiveCount = 1
                decreaseConsecutiveCount = 1
            }

            longestMonotonicSubarrayLength = max(
                intArrayOf(longestMonotonicSubarrayLength, increaseConsecutiveCount, decreaseConsecutiveCount)
            )
        }

        return longestMonotonicSubarrayLength
    }
}

//// NOTE: Naive Solution
//class Solution {
//    fun getMaximumMonotonousSubarrayLength(nums: IntArray): Int {
//        var maxMonotonousSubarrayLength = 1
//        var currentMonotonousSubarrayLength = 1
//
//        for (i in 1 until nums.size) {
//            if (nums[i] > nums[i - 1]) {
//                currentMonotonousSubarrayLength++
//            } else {
//                maxMonotonousSubarrayLength = max(
//                    maxMonotonousSubarrayLength, currentMonotonousSubarrayLength
//                )
//                currentMonotonousSubarrayLength = 1
//            }
//        }
//
//        return max(maxMonotonousSubarrayLength, currentMonotonousSubarrayLength)
//    }
//
//    fun reverseArray(nums: IntArray): IntArray {
//        val reversedArray = IntArray(nums.size)
//
//        for (i in nums.indices) {
//            reversedArray[i] = nums[nums.size - i - 1]
//        }
//
//        return reversedArray
//    }
//
//    fun max(a: Int, b: Int): Int {
//        return if (a > b) a else b
//    }
//
//    fun longestMonotonicSubarray(nums: IntArray): Int {
//        return max(
//            getMaximumMonotonousSubarrayLength(nums), getMaximumMonotonousSubarrayLength(
//                reverseArray(nums),
//            )
//        )
//    }
//}

File: leetcode/easy/3105-longest-strictly-increasing-or-strictly-decreasing-subarray/3105-longest-strictly-increasing-or-strictly-decreasing-subarray.ts
// 3110-score-of-a-string
// URL: https://leetcode.com/problems/score-of-a-string/description/

// # NOTE: top python solution of leetcode + modified by me
function longestMonotonicSubarray(nums: number[]): number {
    let maxConsecutiveCount = 1;
    let increasingConesecutiveCount = 1;
    let decreasingConsecutiveCount = 1;
  
    for (let i = 0; i < nums.length - 1; i++) {
      if (nums[i] < nums[i + 1]) {
        increasingConesecutiveCount += 1;
        decreasingConsecutiveCount = 1;
      } else if (nums[i] > nums[i + 1]) {
        decreasingConsecutiveCount += 1;
        increasingConesecutiveCount = 1;
      } else {
        increasingConesecutiveCount = 1;
        decreasingConsecutiveCount = 1;
      }
  
      maxConsecutiveCount = Math.max(
        maxConsecutiveCount,
        increasingConesecutiveCount,
        decreasingConsecutiveCount,
      );
    }
  
    return maxConsecutiveCount;
  }
  
  
  // NOTE: naive O(2n)
  // function getLongestMonotonicSubarrayLength(nums: number[]): number {
  //   let max = 0;
  //   let current = 0;
  //   let prev = nums[0];
  //
  //   for (let i = 1; i < nums.length; i++) {
  //     if (nums[i] > prev) {
  //       current++;
  //     } else {
  //       max = Math.max(max, current);
  //       current = 0;
  //     }
  //     prev = nums[i];
  //   }
  //
  //   return Math.max(max, current) + 1;
  // }
  //
  // function longestMonotonicSubarray(nums: number[]): number {
  //   return Math.max(
  //     getLongestMonotonicSubarrayLength(nums),
  //     getLongestMonotonicSubarrayLength(nums.reverse())
  //   );
  // };
  